File: dsa/li15_dp/Partitions_L887_SuperEggDrop.java

public class Partitions_L887_SuperEggDrop {
	// Problem: https://leetcode.com/problems/super-egg-drop/
	public int superEggDrop(int eggs, int floors){
        int dp[][] = new int[eggs+1][floors+1];
        for(int i=0; i <= floors; i++)
            dp[1][i] = i;

        for(int e = 2; e <= eggs; e++){
            for(int f = 1; f <=floors; f++){
                int min = Integer.MAX_VALUE;
                for(int k = 1; k<=f ; k++){
                    int moves = 1 + Math.max(dp[e-1][k-1], dp[e][f-k]);
                    if(moves < min){
                        min = moves;
                    }
                }
                dp[e][f] = min;
            }
        }
        return dp[eggs][floors];
    }

    public int calculateRecursive(int eggs, int floors){
        if(eggs == 1)
            return floors;
        if(floors == 0)
            return 0;

        int min = (int)1e9;
        for(int k=1; k <= floors; k++){
            int moves = 1 + Math.max(calculateRecursive(eggs-1, k-1), calculateRecursive(eggs, floors-k));
            if(moves < min){
                min = moves;
            }
        }
        return min;
    }

    public static void main(String args[]){
        Partitions_L887_SuperEggDrop ed = new Partitions_L887_SuperEggDrop();

        int r = ed.superEggDrop(2,6);
        System.out.println(r);
    }

}

File: dsa/patterns/15_dp/python/partitions_l887_super_egg_drop.py
# DP / Partition DP
#
# Problem: https://leetcode.com/problems/super-egg-drop/

def superEggDrop(eggs, floors):
    dp = [[0] * (floors + 1) for _ in range(eggs + 1)]

    # Base case: only 1 egg
    for i in range(floors + 1):
        dp[1][i] = i

    for e in range(2, eggs + 1):
        for f in range(1, floors + 1):
            minimum = float('inf')
            for k in range(1, f + 1):
                moves = 1 + max(dp[e - 1][k - 1], dp[e][f - k])
                minimum = min(minimum, moves)
            dp[e][f] = minimum

    return dp[eggs][floors]


def calculateRecursive(eggs, floors):
    if eggs == 1:
        return floors
    if floors == 0:
        return 0

    minimum = float('inf')
    for k in range(1, floors + 1):
        moves = 1 + max(
            calculateRecursive(eggs - 1, k - 1),
            calculateRecursive(eggs, floors - k)
        )
        minimum = min(minimum, moves)

    return minimum


# -------- Test --------
print(superEggDrop(2, 6))        # Expected: 3
# print(calculateRecursive(2, 6))  # Also returns 3 (but very slow)
